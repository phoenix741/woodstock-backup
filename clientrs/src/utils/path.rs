use globset::{GlobBuilder, GlobSetBuilder};
use percent_encoding::{percent_decode_str, utf8_percent_encode, NON_ALPHANUMERIC};
use std::{
    ffi::{OsStr, OsString},
    path::{Path, PathBuf},
};

/// Converts a vector of byte vectors to a vector of string slices.
///   
/// # Arguments
///
/// * `vec` - A vector of byte vectors.
///
/// # Returns
///
/// A vector of string slices.
///
#[must_use]
pub fn vec_to_str(vec: &Vec<String>) -> Vec<&str> {
    let mut vec_of_str: Vec<&str> = Vec::new();

    for value in vec {
        vec_of_str.push(value);
    }

    vec_of_str
}

/// Converts a path to a vector of bytes.
///
/// # Arguments
///
/// * `path` - A path.
///
/// # Returns
///
/// A vector of bytes.
///
#[must_use]
pub fn osstr_to_vec(path: &OsStr) -> Vec<u8> {
    path.as_encoded_bytes().to_vec()
}

/// Converts a vector of bytes to a `PathBuf`.
///
/// # Arguments
///
/// * `vec` - A vector of bytes.
///
/// # Returns
///
/// A `PathBuf`.
///
#[must_use]
pub fn vec_to_osstr(vec: &[u8]) -> OsString {
    unsafe { OsString::from_encoded_bytes_unchecked(vec.to_owned()) }
}

/// Converts a path to a vector of bytes.
///
/// # Arguments
///
/// * `path` - A path.
///
/// # Returns
///
/// A vector of bytes.
///
#[must_use]
pub fn path_to_vec(path: &Path) -> Vec<u8> {
    osstr_to_vec(path.as_os_str())
}

/// Converts a vector of bytes to a `PathBuf`.
///
/// # Arguments
///
/// * `vec` - A vector of bytes.
///
/// # Returns
///
/// A `PathBuf`.
///
#[must_use]
pub fn vec_to_path(vec: &[u8]) -> PathBuf {
    PathBuf::from(vec_to_osstr(vec))
}

/// Converts a list of string slices to a `GlobSet`.
///
/// # Arguments
///
/// * `list` - A list of string slices.
///
/// # Returns
///
/// A `Result` containing the `GlobSet` if successful, or an error if the pattern cannot be parsed.
///
/// # Errors
///
/// An error is returned if the pattern cannot be parsed.
///
pub fn list_to_globset(list: &[&str]) -> Result<globset::GlobSet, globset::Error> {
    let mut builder = GlobSetBuilder::new();
    for pattern in list {
        builder.add(GlobBuilder::new(pattern).build()?);
    }
    builder.build()
}

/// Take a path and mangle it by replacing special characters like the
/// method encodeURIComponent will do in javascript.
///
/// # Arguments
///
/// * `path` - A string slice representing the path to mangle.
///
/// # Returns
///
/// A string slice representing the mangled path.
///
#[must_use]
pub fn mangle(path: &str) -> String {
    utf8_percent_encode(path, NON_ALPHANUMERIC).to_string()
}

/// Take a mangled path and unmangle it by decoding percent-encoded characters.
///
/// # Arguments
///
/// * `path` - A string slice representing the mangled path.
///
/// # Returns
///
/// A string slice representing the unmangled path.
///
#[must_use]
pub fn unmangle(path: &str) -> String {
    percent_decode_str(path).decode_utf8_lossy().to_string()
}
