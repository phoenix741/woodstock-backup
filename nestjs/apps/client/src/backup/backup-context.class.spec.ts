import { NotFoundException } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import {
  FileBrowserService,
  FileReader,
  IndexManifest,
  Manifest,
  ManifestService,
  ProtobufService,
} from '@woodstock/shared';
import { count, from, toArray } from 'ix/asynciterable';
import * as Long from 'long';
import { join } from 'path';
import { BackupContext } from './backup-context.class';

describe('BackupContext', () => {
  let service: BackupContext;
  let manifestService: ManifestService;
  let fileReader: FileReader;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ManifestService, ProtobufService, FileReader, FileBrowserService],
    }).compile();

    manifestService = module.get<ManifestService>(ManifestService);
    fileReader = module.get<FileReader>(FileReader);
    service = new BackupContext(fileReader, manifestService);
  });

  describe('#RefreshCache', () => {
    test('#RefreshCache should create one manifest from scratch ', async () => {
      let cnt = 0;
      jest.spyOn(manifestService, 'deleteManifest').mockResolvedValue(undefined);
      jest.spyOn(manifestService, 'writeJournalEntry').mockImplementation(async (entries) => {
        cnt = await count(entries);
      });

      function* myIterator() {
        yield { header: { sharePath: Buffer.from('/home') } };
        yield { fileManifest: { path: Buffer.from('/test1'), stats: { size: Long.fromNumber(100) } } };
        yield { fileManifest: { path: Buffer.from('/test2'), stats: { size: Long.fromNumber(200) } } };
        yield { fileManifest: { path: Buffer.from('/test3'), stats: { size: Long.fromNumber(300) } } };
        yield { fileManifest: { path: Buffer.from('/test4'), stats: { size: Long.fromNumber(400) } } };
        yield { fileManifest: { path: Buffer.from('/test5'), stats: { size: Long.fromNumber(500) } } };
        yield { fileManifest: { path: Buffer.from('/test6'), stats: { size: Long.fromNumber(600) } } };
      }

      const reply = await service.refreshCache(from(myIterator()));

      expect(manifestService.deleteManifest).toHaveBeenCalledTimes(1);
      expect(manifestService.deleteManifest).toHaveBeenCalledWith(new Manifest('backups.%2Fhome', '/tmp'));
      expect(cnt).toBe(6);
      expect(reply).toMatchSnapshot('RefreshCacheReply');
    });

    test('#RefreshCache should create two manifest from scratch ', async () => {
      const cnt: Record<string, number> = {};
      jest.spyOn(manifestService, 'deleteManifest').mockResolvedValue(undefined);
      jest.spyOn(manifestService, 'writeJournalEntry').mockImplementation(async (entries, manifest: Manifest) => {
        cnt[manifest.journalPath] = await count(entries);
      });

      function* myIterator() {
        yield { header: { sharePath: Buffer.from('/home') } };
        yield { fileManifest: { path: Buffer.from('/test1'), stats: { size: Long.fromNumber(100) } } };
        yield { fileManifest: { path: Buffer.from('/test2'), stats: { size: Long.fromNumber(200) } } };
        yield { fileManifest: { path: Buffer.from('/test3'), stats: { size: Long.fromNumber(300) } } };
        yield { fileManifest: { path: Buffer.from('/test4'), stats: { size: Long.fromNumber(400) } } };
        yield { fileManifest: { path: Buffer.from('/test5'), stats: { size: Long.fromNumber(500) } } };
        yield { fileManifest: { path: Buffer.from('/test6'), stats: { size: Long.fromNumber(600) } } };
        yield {
          header: { sharePath: Buffer.from('/etc') },
          fileManifest: { path: Buffer.from('/test32'), stats: { size: Long.fromNumber(100) } },
        };
        yield { fileManifest: { path: Buffer.from('/test7'), stats: { size: Long.fromNumber(100) } } };
        yield { fileManifest: { path: Buffer.from('/test8'), stats: { size: Long.fromNumber(200) } } };
        yield { fileManifest: { path: Buffer.from('/test9'), stats: { size: Long.fromNumber(300) } } };
      }

      const reply = await service.refreshCache(from(myIterator()));

      expect(manifestService.deleteManifest).toHaveBeenCalledTimes(2);
      expect(manifestService.deleteManifest).toHaveBeenCalledWith(new Manifest('backups.%2Fhome', '/tmp'));
      expect(cnt).toMatchInlineSnapshot(`
              Object {
                "/tmp/backups.%2Fetc.journal": 4,
                "/tmp/backups.%2Fhome.journal": 6,
              }
          `);
      expect(reply).toMatchSnapshot('RefreshCacheReply');
    });

    test('#RefreshCache should create error', async () => {
      const cnt: Record<string, number> = {};
      jest.spyOn(manifestService, 'deleteManifest').mockResolvedValue(undefined);
      jest.spyOn(manifestService, 'writeJournalEntry').mockImplementation(async (entries, manifest: Manifest) => {
        cnt[manifest.journalPath] = await count(entries);
      });

      function* myIterator() {
        yield { header: { sharePath: Buffer.from('/home') } };
        yield { fileManifest: { path: Buffer.from('/test1'), stats: { size: Long.fromNumber(100) } } };
        throw new NotFoundException('Should be return error');
      }

      const reply = await service.refreshCache(from(myIterator()));

      expect(manifestService.deleteManifest).toHaveBeenCalledTimes(1);
      expect(manifestService.deleteManifest).toHaveBeenCalledWith(new Manifest('backups.%2Fhome', '/tmp'));
      expect(cnt).toMatchInlineSnapshot(`Object {}`);
      expect(reply).toMatchSnapshot('RefreshCacheReply');
    });
  });

  describe('LaunchBackup', () => {
    let mockIndex: IndexManifest;

    beforeEach(() => {
      mockIndex = new IndexManifest();
      jest.spyOn(manifestService, 'loadIndex').mockImplementation(async (manifest: Manifest) => {
        expect(manifest.manifestPath).toBe('/tmp/backups.%2Fhome%2Fphoenix%2FDownloads%2Ftest.manifest');
        return mockIndex;
      });
      jest.spyOn(fileReader, 'getFiles').mockImplementation((index: IndexManifest) =>
        from(
          (function* fileReaderGetFiles(index: IndexManifest) {
            expect(index).toBe(mockIndex);
            const entry = index.getEntry(Buffer.from('/42093_X_Hot Rod.pdf'));
            if (entry) {
              entry.markViewed = true;
            }
            yield {
              chunks: [
                Buffer.from('782199a8de21390a804df2db8cc308c4d8966c10e181a252f96af35c51e9682f', 'hex'),
                Buffer.from('4a8602e2bc9f3eb9f744da31b0e1abf5474023e94a8bba97ecf544f578372fff', 'hex'),
              ],
              path: Buffer.from('/42093_X_Hot Rod.pdf'),
              sha256: Buffer.from('46b475c703c639f21c725daf1eed193ab1ea66a069028b512e0bca60533e037c', 'hex'),
              stats: {
                created: new Long(1271037455, 376),
                groupId: new Long(1000, 0),
                lastModified: new Long(1656647184, 371),
                lastRead: new Long(1063485888, 372),
                mode: new Long(33188, 0),
                ownerId: new Long(1000, 0),
                size: new Long(4503209, 0),
              },
            };
          })(index),
        ),
      );
    });

    test('should make backup', async () => {
      const share = {
        sharePath: Buffer.from('/home/phoenix/Downloads/test'),
        includes: [Buffer.from('*.pdf')],
        excludes: [Buffer.from('*.bak')],
      };
      const it = await toArray(service.launchBackup(share));
      expect(it).toMatchSnapshot('Launch backup (normal)');
    });

    test('should make backup with index (remove elements)', async () => {
      mockIndex.add({
        path: Buffer.from('/42093_X_Hot Rod.pdf'),
        stats: {},
        acl: [],
        xattr: {},
        chunks: [],
      });
      mockIndex.add({
        path: Buffer.from('/fichier supprimÃ©.pdf'),
        stats: {},
        acl: [],
        xattr: {},
        chunks: [],
      });
      const share = {
        sharePath: Buffer.from('/home/phoenix/Downloads/test'),
        includes: [],
        excludes: [],
      };
      const it = await toArray(service.launchBackup(share));
      expect(it).toMatchSnapshot('Launch backup (normal)');
    });

    test('but return an error', async () => {
      jest.spyOn(manifestService, 'loadIndex').mockRejectedValue(new Error("Can't read files"));

      const share = {
        sharePath: Buffer.from('/home/phoenix/Downloads/test'),
        includes: [],
        excludes: [],
      };
      const it = await toArray(service.launchBackup(share));
      expect(it).toMatchSnapshot('Launch backup (normal)');
    });
  });

  describe('#getChunk', () => {
    test('download a file', async () => {
      const testFile = join(__dirname, 'backup-context.class.ts');
      const chunks = service.getChunk({
        filename: Buffer.from(testFile),
        position: Long.fromNumber(0),
        size: Long.fromNumber(100000),
        sha256: Buffer.from('922d7c9eec991ad5eee982292fcb942715be283e4f5f2e18453602f11832ec7f', 'hex'),
        failIfWrongHash: true,
      });

      const allChunks = await toArray(chunks);

      expect(allChunks.map((c) => c.data.toString())).toMatchSnapshot('getChunk');
    });

    test('download a file, wrong hash', async () => {
      const testFile = join(__dirname, 'backup-context.class.ts');
      const chunks = service.getChunk({
        filename: Buffer.from(testFile),
        position: Long.fromNumber(0),
        size: Long.fromNumber(100000),
        sha256: Buffer.from('13', 'hex'),
        failIfWrongHash: true,
      });

      await expect(toArray(chunks)).rejects.toThrow(/The chunk .* should have a sha of 13, but is .*/);
    });

    test('download a file, piece of file', async () => {
      const testFile = join(__dirname, 'backup-context.class.ts');
      const chunks = service.getChunk({
        filename: Buffer.from(testFile),
        position: Long.fromNumber(10),
        size: Long.fromNumber(30),
        sha256: Buffer.from('bbc1eaf9e6f958bfb798e0c572d295ed53b5d4ab9fa7cc2e0d8c71efdc8b090a', 'hex'),
        failIfWrongHash: true,
      });

      const allChunks = await toArray(chunks);

      expect(allChunks.map((c) => c.data.toString())).toMatchSnapshot('getChunk');
    });
  });
});
