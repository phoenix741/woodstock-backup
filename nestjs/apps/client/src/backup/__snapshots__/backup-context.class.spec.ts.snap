// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackupContext #RefreshCache #RefreshCache should create error: RefreshCacheReply 1`] = `
Object {
  "code": 1,
  "message": "Should be return error",
}
`;

exports[`BackupContext #RefreshCache #RefreshCache should create error: deleteManifest 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Manifest {
        "fileListPath": "/tmp/backups.%2Fhome.filelist",
        "journalPath": "/tmp/backups.%2Fhome.journal",
        "lockPath": "/tmp/backups.%2Fhome.lock",
        "manifestPath": "/tmp/backups.%2Fhome.manifest",
        "newPath": "/tmp/backups.%2Fhome.new",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`BackupContext #RefreshCache #RefreshCache should create one manifest from scratch : RefreshCacheReply 1`] = `
Object {
  "code": 0,
}
`;

exports[`BackupContext #RefreshCache #RefreshCache should create one manifest from scratch : deleteManifest 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Manifest {
        "fileListPath": "/tmp/backups.%2Fhome.filelist",
        "journalPath": "/tmp/backups.%2Fhome.journal",
        "lockPath": "/tmp/backups.%2Fhome.lock",
        "manifestPath": "/tmp/backups.%2Fhome.manifest",
        "newPath": "/tmp/backups.%2Fhome.new",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`BackupContext #RefreshCache #RefreshCache should create two manifest from scratch : RefreshCacheReply 1`] = `
Object {
  "code": 0,
}
`;

exports[`BackupContext #RefreshCache #RefreshCache should create two manifest from scratch : deleteManifest 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Manifest {
        "fileListPath": "/tmp/backups.%2Fhome.filelist",
        "journalPath": "/tmp/backups.%2Fhome.journal",
        "lockPath": "/tmp/backups.%2Fhome.lock",
        "manifestPath": "/tmp/backups.%2Fhome.manifest",
        "newPath": "/tmp/backups.%2Fhome.new",
      },
    ],
    Array [
      Manifest {
        "fileListPath": "/tmp/backups.%2Fetc.filelist",
        "journalPath": "/tmp/backups.%2Fetc.journal",
        "lockPath": "/tmp/backups.%2Fetc.lock",
        "manifestPath": "/tmp/backups.%2Fetc.manifest",
        "newPath": "/tmp/backups.%2Fetc.new",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`BackupContext #getChunk download a file, piece of file: getChunk 1`] = `
Array [
  "adRequestException, InternalSe",
]
`;

exports[`BackupContext #getChunk download a file: getChunk 1`] = `
Array [
  "import { BadRequestException, InternalServerErrorException, Logger } from '@nestjs/common';
import {
  ChunkInformation,
  FileChunk,
  FileHashReader,
  FileManifest,
  FileReader,
  globStringToRegex,
  LaunchBackupReply,
  mangle,
  Manifest,
  ManifestService,
  notUndefined,
  RefreshCacheReply,
  RefreshCacheRequest,
  Share,
  StatusCode,
} from '@woodstock/shared';
import { createReadStream, ReadStream } from 'fs';
import { throwIfAborted } from 'ix/aborterror';
import { AsyncIterableX, concat, create, from, of, pipe } from 'ix/asynciterable';
import { catchError, concatAll, filter, finalize, map } from 'ix/asynciterable/operators';

export class BackupContext {
  private logger = new Logger(BackupContext.name);

  constructor(private fileReader: FileReader, private manifestService: ManifestService) {}

  private async createManifestForSharePath(sharePath: Buffer, source: AsyncIterable<FileManifest | undefined>) {
    const manifest = new Manifest(\`backups.\${mangle(sharePath)}\`, '/tmp/');
    await this.manifestService.deleteManifest(manifest);

    const entries$ = pipe(
      source,
      notUndefined<FileManifest>(),
      map((fileManifest) => this.manifestService.toAddJournalEntry(fileManifest, true)),
    );
    await this.manifestService.writeJournalEntry(entries$, manifest);

    await this.manifestService.compact(manifest);
  }

  private async createManifestsFromSource(request: AsyncIterableX<RefreshCacheRequest>) {
    // First search the first header
    const it = request[Symbol.asyncIterator]();
    let next: IteratorResult<RefreshCacheRequest, RefreshCacheRequest>;
    let sharePath: Buffer | undefined;
    let firstManifest: FileManifest | undefined;

    while (!(next = await it.next()).done) {
      const { header, fileManifest } = next.value;
      sharePath = header?.sharePath;
      firstManifest = fileManifest;
      if (sharePath) {
        break;
      }
    }
    if (!sharePath) {
      throw new InternalServerErrorException();
    }

    while (!next.done) {
      // Then create an iterator for the rest
      const dataIt = create<FileManifest | undefined>((signal) => {
        return {
          async next() {
            throwIfAborted(signal);

            next = await it.next();
            if (next.done) {
              return { done: true, value: undefined };
            }
            const { header, fileManifest } = next.value;
            if (header?.sharePath) {
              sharePath = header.sharePath;
              firstManifest = fileManifest;
              return { done: true, value: undefined };
            }
            return { value: fileManifest };
          },
        };
      });

      // Then create the manifest
      await this.createManifestForSharePath(sharePath, firstManifest ? concat(of(firstManifest), dataIt) : dataIt);
    }
  }

  async refreshCache(request: AsyncIterableX<RefreshCacheRequest>): Promise<RefreshCacheReply> {
    try {
      await this.createManifestsFromSource(request);

      return { code: StatusCode.Ok };
    } catch (err) {
      return { code: StatusCode.Failed, message: err.message };
    }
  }

  launchBackup(share: Share): AsyncIterableX<LaunchBackupReply> {
    const manifest = new Manifest(\`backups.\${mangle(share.sharePath)}\`, '/tmp/');

    const createIndex$ = from(this.manifestService.loadIndex(manifest));

    const addEntryToIndex$ = createIndex$.pipe(
      map((index) =>
        this.fileReader.getFiles(
          index,
          share.sharePath,
          share.includes?.map((s) => globStringToRegex(s.toString('latin1'))),
          share.excludes?.map((s) => globStringToRegex(s.toString('latin1'))),
        ),
      ),
      concatAll(),
      map((entry) => this.manifestService.toAddJournalEntry(entry, true)),
      map((entry) => ({ entry, response: undefined })),
    );

    const addRemoveToIndex$ = createIndex$.pipe(
      map((index) => from(index.walk())),
      concatAll(),
      filter((entry) => !entry.markViewed),
      map((file) => this.manifestService.toRemoveJournalEntry(file.path)),
      map((entry) => ({ entry, response: undefined })),
    );

    const entries$ = concat(
      addEntryToIndex$,
      addRemoveToIndex$,
      of({ entry: undefined, response: { code: StatusCode.Ok } }),
    );

    return entries$.pipe(
      catchError((err) => {
        return of({
          entry: undefined,
          response: { code: StatusCode.Failed, message: err.message },
        });
      }),
    );
  }

  getChunk(request: ChunkInformation & { failIfWrongHash?: boolean }): AsyncIterableX<FileChunk> {
    const { filename, position, size, sha256 } = request;
    const stream: ReadStream = createReadStream(filename, {
      start: position.toNumber(),
      end: position.add(size).sub(1).toNumber(),
    });
    const hashReader = new FileHashReader();

    stream.pipe(hashReader);
    stream.on('error', (err) => {
      hashReader.emit('error', err);
      hashReader.end();
      stream.close();
    });
    hashReader.on('error', () => {
      hashReader.end();
      stream.close();
    });

    return from(hashReader).pipe(
      map((data) => ({ data })),
      finalize(() => {
        if (request.failIfWrongHash && sha256 && hashReader.hash && !hashReader.hash.equals(sha256)) {
          const message = \`The chunk \${filename.toString()}:\${position}:\${size} should have a sha of \${sha256.toString(
            'hex',
          )}, but is \${hashReader.hash.toString('hex')}\`;
          this.logger.warn(message);

          throw new BadRequestException(message);
        }
      }),
    );
  }
}
",
]
`;

exports[`BackupContext LaunchBackup but return an error: Launch backup (normal) 1`] = `
Array [
  Object {
    "entry": undefined,
    "response": Object {
      "code": 1,
      "message": "Can't read files",
    },
  },
]
`;

exports[`BackupContext LaunchBackup should make backup with index (remove elements): Launch backup (normal) 1`] = `
Array [
  Object {
    "entry": Object {
      "manifest": Object {
        "chunks": Array [
          Object {
            "data": Array [
              120,
              33,
              153,
              168,
              222,
              33,
              57,
              10,
              128,
              77,
              242,
              219,
              140,
              195,
              8,
              196,
              216,
              150,
              108,
              16,
              225,
              129,
              162,
              82,
              249,
              106,
              243,
              92,
              81,
              233,
              104,
              47,
            ],
            "type": "Buffer",
          },
          Object {
            "data": Array [
              74,
              134,
              2,
              226,
              188,
              159,
              62,
              185,
              247,
              68,
              218,
              49,
              176,
              225,
              171,
              245,
              71,
              64,
              35,
              233,
              74,
              139,
              186,
              151,
              236,
              245,
              68,
              245,
              120,
              55,
              47,
              255,
            ],
            "type": "Buffer",
          },
        ],
        "path": Object {
          "data": Array [
            47,
            52,
            50,
            48,
            57,
            51,
            95,
            88,
            95,
            72,
            111,
            116,
            32,
            82,
            111,
            100,
            46,
            112,
            100,
            102,
          ],
          "type": "Buffer",
        },
        "sha256": Object {
          "data": Array [
            70,
            180,
            117,
            199,
            3,
            198,
            57,
            242,
            28,
            114,
            93,
            175,
            30,
            237,
            25,
            58,
            177,
            234,
            102,
            160,
            105,
            2,
            139,
            81,
            46,
            11,
            202,
            96,
            83,
            62,
            3,
            124,
          ],
          "type": "Buffer",
        },
        "stats": Object {
          "created": Long {
            "high": 376,
            "low": 1271037455,
            "unsigned": false,
          },
          "groupId": Long {
            "high": 0,
            "low": 1000,
            "unsigned": false,
          },
          "lastModified": Long {
            "high": 371,
            "low": 1656647184,
            "unsigned": false,
          },
          "lastRead": Long {
            "high": 372,
            "low": 1063485888,
            "unsigned": false,
          },
          "mode": Long {
            "high": 0,
            "low": 33188,
            "unsigned": false,
          },
          "ownerId": Long {
            "high": 0,
            "low": 1000,
            "unsigned": false,
          },
          "size": Long {
            "high": 0,
            "low": 4503209,
            "unsigned": false,
          },
        },
      },
      "type": 0,
    },
    "response": undefined,
  },
  Object {
    "entry": Object {
      "manifest": Object {
        "acl": Array [],
        "chunks": Array [],
        "path": Object {
          "data": Array [
            47,
            102,
            105,
            99,
            104,
            105,
            101,
            114,
            32,
            115,
            117,
            112,
            112,
            114,
            105,
            109,
            195,
            169,
            46,
            112,
            100,
            102,
          ],
          "type": "Buffer",
        },
        "stats": undefined,
        "xattr": Object {},
      },
      "type": 2,
    },
    "response": undefined,
  },
  Object {
    "entry": undefined,
    "response": Object {
      "code": 0,
    },
  },
]
`;

exports[`BackupContext LaunchBackup should make backup: Launch backup (normal) 1`] = `
Array [
  Object {
    "entry": Object {
      "manifest": Object {
        "chunks": Array [
          Object {
            "data": Array [
              120,
              33,
              153,
              168,
              222,
              33,
              57,
              10,
              128,
              77,
              242,
              219,
              140,
              195,
              8,
              196,
              216,
              150,
              108,
              16,
              225,
              129,
              162,
              82,
              249,
              106,
              243,
              92,
              81,
              233,
              104,
              47,
            ],
            "type": "Buffer",
          },
          Object {
            "data": Array [
              74,
              134,
              2,
              226,
              188,
              159,
              62,
              185,
              247,
              68,
              218,
              49,
              176,
              225,
              171,
              245,
              71,
              64,
              35,
              233,
              74,
              139,
              186,
              151,
              236,
              245,
              68,
              245,
              120,
              55,
              47,
              255,
            ],
            "type": "Buffer",
          },
        ],
        "path": Object {
          "data": Array [
            47,
            52,
            50,
            48,
            57,
            51,
            95,
            88,
            95,
            72,
            111,
            116,
            32,
            82,
            111,
            100,
            46,
            112,
            100,
            102,
          ],
          "type": "Buffer",
        },
        "sha256": Object {
          "data": Array [
            70,
            180,
            117,
            199,
            3,
            198,
            57,
            242,
            28,
            114,
            93,
            175,
            30,
            237,
            25,
            58,
            177,
            234,
            102,
            160,
            105,
            2,
            139,
            81,
            46,
            11,
            202,
            96,
            83,
            62,
            3,
            124,
          ],
          "type": "Buffer",
        },
        "stats": Object {
          "created": Long {
            "high": 376,
            "low": 1271037455,
            "unsigned": false,
          },
          "groupId": Long {
            "high": 0,
            "low": 1000,
            "unsigned": false,
          },
          "lastModified": Long {
            "high": 371,
            "low": 1656647184,
            "unsigned": false,
          },
          "lastRead": Long {
            "high": 372,
            "low": 1063485888,
            "unsigned": false,
          },
          "mode": Long {
            "high": 0,
            "low": 33188,
            "unsigned": false,
          },
          "ownerId": Long {
            "high": 0,
            "low": 1000,
            "unsigned": false,
          },
          "size": Long {
            "high": 0,
            "low": 4503209,
            "unsigned": false,
          },
        },
      },
      "type": 0,
    },
    "response": undefined,
  },
  Object {
    "entry": undefined,
    "response": Object {
      "code": 0,
    },
  },
]
`;
